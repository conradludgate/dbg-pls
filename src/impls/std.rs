mod collections;
mod fnptr;
mod tuple;

use std::{
    ops,
    ops::ControlFlow,
    rc::Rc,
    sync::{Arc, Mutex, MutexGuard, TryLockError},
    task::Poll,
};

use crate::{DebugWith, Formatter};
use syn::RangeLimits;
use syn::__private::Span;

impl<W, T: ?Sized + DebugWith<W>> DebugWith<W> for Box<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        DebugWith::fmt(&**self, with, f);
    }
}

impl<W, D: DebugWith<W> + ?Sized> DebugWith<W> for *mut D {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        <*const D>::fmt(&(*self).cast_const(), with, f);
    }
}

impl<W, D: ?Sized> DebugWith<W> for *const D {
    fn fmt(&self, _with: &W, f: Formatter<'_>) {
        /// Since the formatting will be identical for all pointer types, use a non-monomorphized
        /// implementation for the actual formatting to reduce the amount of codegen work needed
        fn inner(ptr: *const (), f: Formatter<'_>) {
            let output = format!("{:#x?}", ptr as usize);
            f.write_expr(syn::ExprLit {
                attrs: vec![],
                lit: syn::LitFloat::new(&output, Span::call_site()).into(),
            });
        }

        inner((*self).cast(), f);
    }
}

impl<'a, W, D: DebugWith<W> + ?Sized> DebugWith<W> for &'a D {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        D::fmt(self, with, f);
    }
}

impl<'a, W, D: DebugWith<W> + ?Sized> DebugWith<W> for &'a mut D {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        D::fmt(self, with, f);
    }
}

impl<W, T: ?Sized + DebugWith<W>> DebugWith<W> for Rc<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        DebugWith::fmt(&**self, with, f);
    }
}

impl<W, T: ?Sized + DebugWith<W>> DebugWith<W> for Arc<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        DebugWith::fmt(&**self, with, f);
    }
}

impl<W, T: ?Sized + DebugWith<W>> DebugWith<W> for MutexGuard<'_, T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        DebugWith::fmt(&**self, with, f);
    }
}

impl<W, T: ?Sized + DebugWith<W>> DebugWith<W> for Mutex<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        let d = f.debug_struct("Mutex");
        match self.try_lock() {
            Ok(guard) => d.field_with("data", &&*guard, with),
            Err(TryLockError::Poisoned(err)) => d.field_with("data", &&**err.get_ref(), with),
            Err(TryLockError::WouldBlock) => d.field("data", &"<locked>"),
        }
        .field("poisoned", &self.is_poisoned())
        .finish_non_exhaustive();
    }
}

macro_rules! debug_integers {
    ($($T:ident)*) => {$(
        impl<W> DebugWith<W> for $T {
            fn fmt(&self, _with: &W, f: Formatter<'_>) {
                let mut buf = itoa::Buffer::new();
                f.write_expr(syn::ExprLit {
                    attrs: vec![],
                    lit: syn::LitInt::new(buf.format(*self), Span::call_site()).into(),
                });
            }
        }
    )*};
}

debug_integers! {
  i8 i16 i32 i64 i128 isize
  u8 u16 u32 u64 u128 usize
}

macro_rules! debug_non_zero_integers {
    ($($T:ident)*) => {$(
        impl<W> DebugWith<W> for std::num::$T {
            fn fmt(&self, _with: &W, f: Formatter<'_>) {
                let mut buf = itoa::Buffer::new();
                f.write_expr(syn::ExprLit {
                    attrs: vec![],
                    lit: syn::LitInt::new(buf.format(self.get()), Span::call_site()).into(),
                });
            }
        }
    )*};
}

debug_non_zero_integers! {
  NonZeroI8 NonZeroI16 NonZeroI32 NonZeroI64 NonZeroI128 NonZeroIsize
  NonZeroU8 NonZeroU16 NonZeroU32 NonZeroU64 NonZeroU128 NonZeroUsize
}

macro_rules! debug_floats {
    ($ty:ident) => {
        impl<W> DebugWith<W> for $ty {
            fn fmt(&self, _with: &W, f: Formatter<'_>) {
                let mut buf = ryu::Buffer::new();
                f.write_expr(syn::ExprLit {
                    attrs: vec![],
                    lit: syn::LitFloat::new(buf.format(*self), Span::call_site()).into(),
                });
            }
        }
    };
}

debug_floats! { f32 }
debug_floats! { f64 }

impl<W> DebugWith<W> for bool {
    fn fmt(&self, _with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprLit {
            attrs: vec![],
            lit: syn::Lit::Bool(syn::LitBool {
                value: *self,
                span: Span::call_site(),
            }),
        });
    }
}

impl<W, D: DebugWith<W>> DebugWith<W> for [D] {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        f.debug_list().entries_with(self, with).finish();
    }
}

impl<W, D: DebugWith<W>, const N: usize> DebugWith<W> for [D; N] {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        f.debug_list().entries_with(self, with).finish();
    }
}

impl<W> DebugWith<W> for char {
    fn fmt(&self, _with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprLit {
            attrs: vec![],
            lit: syn::LitChar::new(*self, Span::call_site()).into(),
        });
    }
}

impl<W> DebugWith<W> for str {
    fn fmt(&self, _with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprLit {
            attrs: vec![],
            lit: syn::LitStr::new(self, Span::call_site()).into(),
        });
    }
}

impl<W> DebugWith<W> for String {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        DebugWith::fmt(self.as_str(), with, f);
    }
}

impl<W, T: DebugWith<W>, E: DebugWith<W>> DebugWith<W> for Result<T, E> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        match self {
            Ok(t) => f.debug_tuple_struct("Ok").field_with(t, with).finish(),
            Err(e) => f.debug_tuple_struct("Err").field_with(e, with).finish(),
        }
    }
}

impl<W, B: DebugWith<W>, C: DebugWith<W>> DebugWith<W> for ControlFlow<B, C> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        match self {
            ControlFlow::Break(b) => f.debug_tuple_struct("Break").field_with(b, with).finish(),
            ControlFlow::Continue(c) => f
                .debug_tuple_struct("Continue")
                .field_with(c, with)
                .finish(),
        }
    }
}

impl<W, T: DebugWith<W>> DebugWith<W> for Option<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        match self {
            Some(t) => f.debug_tuple_struct("Some").field_with(t, with).finish(),
            None => f.debug_ident("None"),
        }
    }
}

impl<W, T: DebugWith<W>> DebugWith<W> for Poll<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        match self {
            Poll::Ready(t) => f.debug_tuple_struct("Ready").field_with(t, with).finish(),
            Poll::Pending => f.debug_ident("Pending"),
        }
    }
}

impl<W, T: DebugWith<W>> DebugWith<W> for ops::Range<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprRange {
            attrs: vec![],
            start: Some(Box::new(Formatter::process_with(&self.start, with))),
            limits: RangeLimits::HalfOpen(syn::token::DotDot::default()),
            end: Some(Box::new(Formatter::process_with(&self.end, with))),
        });
    }
}

impl<W, T: DebugWith<W>> DebugWith<W> for ops::RangeFrom<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprRange {
            attrs: vec![],
            start: Some(Box::new(Formatter::process_with(&self.start, with))),
            limits: RangeLimits::HalfOpen(syn::token::DotDot::default()),
            end: None,
        });
    }
}

impl<W, T: DebugWith<W>> DebugWith<W> for ops::RangeTo<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprRange {
            attrs: vec![],
            start: None,
            limits: RangeLimits::HalfOpen(syn::token::DotDot::default()),
            end: Some(Box::new(Formatter::process_with(&self.end, with))),
        });
    }
}

impl<W> DebugWith<W> for ops::RangeFull {
    fn fmt(&self, _with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprRange {
            attrs: vec![],
            start: None,
            limits: RangeLimits::HalfOpen(syn::token::DotDot::default()),
            end: None,
        });
    }
}

impl<W, T: DebugWith<W>> DebugWith<W> for ops::RangeInclusive<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprRange {
            attrs: vec![],
            start: Some(Box::new(Formatter::process_with(&self.start(), with))),
            limits: RangeLimits::Closed(syn::token::DotDotEq::default()),
            end: Some(Box::new(Formatter::process_with(&self.end(), with))),
        });
    }
}

impl<W, T: DebugWith<W>> DebugWith<W> for ops::RangeToInclusive<T> {
    fn fmt(&self, with: &W, f: Formatter<'_>) {
        f.write_expr(syn::ExprRange {
            attrs: vec![],
            start: None,
            limits: RangeLimits::Closed(syn::token::DotDotEq::default()),
            end: Some(Box::new(Formatter::process_with(&self.end, with))),
        });
    }
}
