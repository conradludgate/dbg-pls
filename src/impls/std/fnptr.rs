use crate::{DebugPls, Formatter};

macro_rules! fnptr_impls_safety_abi {
    ($FnTy: ty, $($Arg: ident),*) => {
        impl<Ret, $($Arg),*> DebugPls for $FnTy {
            fn fmt(&self, f: Formatter<'_>) {
                // HACK: The intermediate cast as usize is required for AVR
                // so that the address space of the source function pointer
                // is preserved in the final function pointer.
                //
                // https://github.com/avr-rust/rust/issues/143
                DebugPls::fmt(&(*self as usize as *const ()), f)
            }
        }
    }
}

macro_rules! peel {
    ($name:ident, $($other:ident,)*) => (fnptr_impls_args! { $($other),* })
}

macro_rules! fnptr_impls_args {
    ($($Arg: ident),+) => {
        fnptr_impls_safety_abi! { extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
        fnptr_impls_safety_abi! { extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
        fnptr_impls_safety_abi! { extern "C" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
        fnptr_impls_safety_abi! { unsafe extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
        fnptr_impls_safety_abi! { unsafe extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
        fnptr_impls_safety_abi! { unsafe extern "C" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }

        peel! { $($Arg,)+ }
    };
    () => {
        // No variadic functions with 0 parameters
        fnptr_impls_safety_abi! { extern "Rust" fn() -> Ret, }
        fnptr_impls_safety_abi! { extern "C" fn() -> Ret, }
        fnptr_impls_safety_abi! { unsafe extern "Rust" fn() -> Ret, }
        fnptr_impls_safety_abi! { unsafe extern "C" fn() -> Ret, }
    };
}

fnptr_impls_args! { A, B, C, D, E, F, G, H, I, J, K, L }
