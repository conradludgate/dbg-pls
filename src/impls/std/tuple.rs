use crate::{DebugPls, Formatter};

macro_rules! peel {
    ($name:ident, $($other:ident,)*) => (tuple! { $($other,)* })
}

macro_rules! tuple {
    () => ();
    ( $($name:ident,)* ) => (
        impl<$($name:DebugPls),+> DebugPls for ($($name,)+) where last_type!($($name,)+): ?Sized {
            #[allow(non_snake_case, unused_assignments)]
            fn fmt(&self, f: Formatter<'_>) {
                let ($(ref $name,)+) = *self;
                f.debug_tuple()
                $(
                    .field(&$name)
                )+
                    .finish();
            }
        }
        peel! { $($name,)+ }
    )
}

macro_rules! last_type {
    ($a:ident,) => { $a };
    ($a:ident, $($rest_a:ident,)+) => { last_type!($($rest_a,)+) };
}

tuple! { T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, }

impl DebugPls for () {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_tuple().finish();
    }
}
