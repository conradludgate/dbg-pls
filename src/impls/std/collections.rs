use std::collections::{BTreeMap, BTreeSet, BinaryHeap, HashMap, HashSet, LinkedList, VecDeque};

use crate::{DebugPls, Formatter};

impl<K: DebugPls, V: DebugPls, S: ::std::hash::BuildHasher> DebugPls for HashMap<K, V, S> {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_map().entries(self).finish();
    }
}

impl<K: DebugPls, V: DebugPls> DebugPls for BTreeMap<K, V> {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_map().entries(self).finish();
    }
}

impl<V: DebugPls, S: ::std::hash::BuildHasher> DebugPls for HashSet<V, S> {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_set().entries(self).finish();
    }
}

impl<V: DebugPls> DebugPls for BTreeSet<V> {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_set().entries(self).finish();
    }
}

impl<D: DebugPls> DebugPls for Vec<D> {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_list().entries(self).finish();
    }
}

impl<D: DebugPls> DebugPls for VecDeque<D> {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_list().entries(self).finish();
    }
}

impl<D: DebugPls> DebugPls for LinkedList<D> {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_list().entries(self).finish();
    }
}

impl<D: DebugPls> DebugPls for BinaryHeap<D> {
    fn fmt(&self, f: Formatter<'_>) {
        f.debug_list().entries(self).finish();
    }
}
