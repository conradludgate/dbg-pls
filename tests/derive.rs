#![allow(clippy::disallowed_names)]

use std::{
    collections::{BTreeMap, BTreeSet},
    ops,
};

use dbg_pls::DebugPls;

macro_rules! assert_pretty_snapshot {
    ($expr:expr) => {
        let expr = $expr;
        let format = format!(
            ">>> pretty\n{}\n\n>>> color\n{}\n",
            dbg_pls::pretty(&expr),
            dbg_pls::color(&expr)
        );
        insta::assert_snapshot!(insta::_macro_support::AutoName, &format, stringify!($expr));
    };
}

#[derive(DebugPls, Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Demo {
    foo: i32,
    bar: &'static str,
}

#[test]
fn debug_struct() {
    assert_pretty_snapshot!(Demo {
        foo: 5,
        bar: "hello",
    });
}

#[test]
fn debug_struct_big() {
    assert_pretty_snapshot!(Demo {
        foo: 5,
        bar: "Hello, world! I am a very long string",
    });
}

#[test]
fn debug_nested_struct() {
    assert_pretty_snapshot!({
        let mut val = [Demo {
            foo: 5,
            bar: "hello",
        }; 10];
        val[6].bar = "Hello, world! I am a very long string";
        val
    });
}

#[test]
fn debug_small_set() {
    assert_pretty_snapshot!(BTreeSet::from([420, 69]));
}

#[test]
fn debug_nested_set() {
    assert_pretty_snapshot!(BTreeSet::from([
        Demo {
            foo: 5,
            bar: "hello",
        },
        Demo {
            foo: 5,
            bar: "Hello, world! I am a very long string",
        },
    ]));
}

#[test]
fn debug_map() {
    assert_pretty_snapshot!(BTreeMap::from([
        ("hello", 60),
        ("Hello, world! I am a very long string", 12)
    ]));
}

#[test]
fn debug_nested_map() {
    assert_pretty_snapshot!(BTreeMap::from([
        (
            Demo {
                foo: 5,
                bar: "hello",
            },
            60,
        ),
        (
            Demo {
                foo: 5,
                bar: "Hello, world! I am a very long string",
            },
            12,
        ),
    ]));
}

#[derive(DebugPls)]
pub struct Generic<T> {
    arg: T,
}

#[test]
fn debug_generic() {
    assert_pretty_snapshot!(Generic { arg: "string" });
}

mod debug_enum_generic {
    use dbg_pls::DebugPls;

    #[derive(DebugPls)]
    pub enum Option2<T> {
        Some(T),
        None,
        Wtf { foo: i32 },
    }

    #[test]
    fn some() {
        assert_pretty_snapshot!(Option2::Some(42));
    }
    #[test]
    fn none() {
        assert_pretty_snapshot!(Option2::<i32>::None);
    }
    #[test]
    fn wtf() {
        assert_pretty_snapshot!(Option2::<i32>::Wtf { foo: 42 });
    }
}

#[derive(DebugPls)]
pub struct LinkedList {
    value: i32,
    next: Option<Box<LinkedList>>,
}

#[test]
fn debug_recursive() {
    assert_pretty_snapshot!(LinkedList {
        value: 0_i32,
        next: Some(Box::new(LinkedList {
            value: 1_i32,
            next: None,
        })),
    });
}

#[derive(DebugPls)]
pub struct LinkedList2<T> {
    value: T,
    next: Option<Box<LinkedList2<T>>>,
}

#[test]
fn debug_recursive2() {
    assert_pretty_snapshot!(LinkedList2 {
        value: 0_i32,
        next: Some(Box::new(LinkedList2 {
            value: 1_i32,
            next: None,
        })),
    });
}

#[derive(DebugPls)]
pub struct Rangeful<T> {
    range: ops::Range<T>,
    range_from: ops::RangeFrom<T>,
    range_to: ops::RangeTo<T>,
    range_full: ops::RangeFull,
    range_inclusive: ops::RangeInclusive<T>,
    range_inclusive_to: ops::RangeToInclusive<T>,
}

#[test]
fn ranges() {
    assert_pretty_snapshot!(Rangeful {
        range: 1..4,
        range_from: 1..,
        range_to: ..7,
        range_full: ..,
        range_inclusive: 1234..=1236,
        range_inclusive_to: ..=70,
    });
}

#[derive(DebugPls)]
enum EmptyEnum {}
