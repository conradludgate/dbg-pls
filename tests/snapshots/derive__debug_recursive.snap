---
source: tests/derive.rs
expression: "LinkedList {\n    value: 0_i32,\n    next: Some(Box::new(LinkedList { value: 1_i32, next: None })),\n}"
---
>>> pretty
LinkedList {
    value: 0,
    next: Some(LinkedList { value: 1, next: None }),
}

>>> color

[32mLinkedList[0m [31m{[0m
    value[31m:[0m 0[31m,[0m
    next[31m:[0m Some[31m([0m[32mLinkedList[0m [31m{[0m value[31m:[0m 1[31m,[0m next[31m:[0m None [31m}[0m[31m)[0m[31m,[0m
[31m}[0m

